README.md
# SDeraser
*SDeraser (formerly known as formatterpi) is an AIRLAB Infrastructure that enables easy formatting of SD cards. The core functionality consists of a python script that automatically detects newly inserted block devices (such as SD-cards or even USB drives), and subesequently erases the device using the ubuntu software "parted". For security reasons, it is important to note that the script is technically only formatting volumes (as opposed to wiping)!  The build is kept screenless and utilizes a button with a bicolor LED as user interaction via GPIO. A full tutorial for setup up can be found amongst [AIR LABs tutorials](https://airlab.itu.dk/category/tutorial/). Below is listed a quick Raspi Changelog of settings and configurations*

![alt text](https://airlab.itu.dk/wp-content/uploads/sites/71/2023/03/eraserpi.png)


## Contents
- Formatter2.py introduces the basic loop checking for an SD card, and erased upon terminal input = y.
- gpio2.py contains testing codes for working with bi-color LED and Button
- Launcher.sh holds the script used by LXDE-PI Autostart to launch SDeraser on boot
- SDeraser.py holds the final script for the SDeraser station

## Raspi Changelog

### Configurations
- (Optional) Changed Keyboard setup to danish
- removed automatic volume popup: filemanager > edit > preferences > volume management > uncheck "show available media types..."
- Enabled interfaces: Remote GPIO and VNC

### Terminal interactions
```sh
$ sudo apt-get update
$ sudo apt-get upgrade
$ sudo apt-get install parted
```

```sh
# Github configuration (Optional)
$ sudo apt-get install git
$ mkdir gitcode
$ cd gitcode
$ git clone https://github.com/airlabitu/SDeraser.git
```

```sh
# Create launcher script, include contents from this directory:
$ sudo nano launcher.sh

# Make script executable:
$ chmod 755 launcher.sh
```

```sh
# Open the global autostart file and copy all lines to clipboard:
$ sudo nano /etc/xdg/lxsession/LXDE-pi/autostart

# Create user-side autostart script and paste all lines from the global one:
$ cd ~/.config/lxsession/LXDE-pi/
$ sudo nano autostart
# Additionally, add the following line to the autostart script, above "Xscreensaver":
@lxterminal -e "/home/pi/launcher.sh"
```


changelog.txt
This logs sums up changes / settings / systemside that has happened on the PI

Creation:
        > sudo apt-get update
        > sudo apt-get upgrade
        > sudo apt-get install parted

    Danish keyboard setup

    removed automatic volume popup: filemanager > edit > preferences > volume management > uncheck "show available media types..." yadayada

13. march
    Enable remote GPIO in raspberry pi configuration > interfaces

    install github:
        > sudo apt-get install git
        > mkdir gitcode
        > cd gitcode
        > git clone https://github.com/airlabitu/SDeraser.git

20. march
    Create launcher.sh in home directory
        > sudo nano launcher.sh

    Copy existing autostart information from global autostart:
        > sudo nano /etc/xdg/lxsession/LXDE-pi/autostart
        (copy all lines to clipboard)


    Create user-side autostart:
        > cd ~/.config/lxsession/LXDE-pi/
        > sudo nano autostart
    
    Include autostart information from global (paste from clipboard)
    Above Xscreensaver add:
        > @lxterminal -e "/home/pi/launcher.sh"

21. march
    Enable VNC

gpio2.py
# !/bin/python
# gpio2.py script to test GPIO in / out
# Introduces functions to change color on bi-color LED (inside button used)
# Thomas Kaufmanas // AIRLAB ITU // spring 2023

import time
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BOARD)
btnLED1 = 37
btnLED2 = 38
btn = 7

GPIO.setup(btnLED1,GPIO.OUT) # set the pin to output
GPIO.setup(btnLED2,GPIO.OUT) # set the pin to output
GPIO.setup(btn, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)


def btnRed():
	"""Sets button colour to red"""
	GPIO.output(btnLED1, GPIO.HIGH)
	GPIO.output(btnLED2, GPIO.LOW)
def btnGreen():
	"""Sets button colour to green"""
	GPIO.output(btnLED1, GPIO.LOW)
	GPIO.output(btnLED2, GPIO.HIGH)
def btnOff():
	"""Sets button colour to off"""
	GPIO.output(btnLED1, GPIO.LOW)
	GPIO.output(btnLED2, GPIO.LOW)


try:
	while True:
		btnRed()
		print("RED")
		time.sleep(0.5)

		btnGreen()
		print("GREEN")
		time.sleep(0.5)

		btnOff()
		print("OFF")
		time.sleep(0.5)

		print(GPIO.input(btn))

		GPIO.wait_for_edge(btn, GPIO.RISING)

except KeyboardInterrupt:
	GPIO.cleanup()

formatter.py
# !/bin/python
# FormatterPi script 1.0
# Thomas Kaufmanas // AIRLAB // spring 2023

import time
import os

print("AIRLAB FormatterPi running...\n")
print("Reading volumes:")
os.system("lsblk")
time.sleep(1)
print("\n")

if input("Format volume /dev/sda? (y/n): ") == "y":
    print("Unmounting all partitions...")
    os.system("sudo umount /dev/sda?")
    time.sleep(0.5)
    print("Constructing partitioning table...")
    os.system("sudo parted -s -a optimal -- /dev/sda mklabel msdos")
    time.sleep(0.5)
    print("partitioning 100% of volume to fat32...")
    os.system("sudo parted -s -a optimal -- /dev/sda mkpart primary fat32 1MiB 100%")
    time.sleep(0.5)
    print("Cleaning up...")
    os.system("sudo parted -s -- /dev/sda align-check optimal 1")
    time.sleep(0.5)
    print("Formatting Volume...")
    os.system("sudo mkfs.vfat -F32 /dev/sda1")
    time.sleep(1)
    print("\nAll good - remove SD card")
else:
    print("quit")
    


formatter2.py
# !/bin/python
# FormatterPi live running script (2.0).
# Searches for newly inserted volumes (blocks). Will request to delete the first block on list, usually sda or sdb. Cannot delete itself "mmcblk"
# Thomas Kaufmanas // AIRLAB ITU // spring 2023

import time
import os
import subprocess

waitstate = 1

def blockCheck():
    """Use subprocess module to read block devices inserted. Returns a list of names"""
    output = subprocess.Popen("lsblk -l -o NAME", shell=True, stdout=subprocess.PIPE).stdout
    output = output.read()
    output = output.decode("utf-8")
    output = output.strip().split("\n")
    result = output[1:]
    #print("array er:", result)
    return result

def wipeSD(blockname):
    """formats an SD card by name of primary blocks (thereby including any partitions under this)"""
    b = blockname
    print("Unmounting all partitions...")
    os.system(f"sudo umount /dev/{b}?")
    time.sleep(0.5)
    print("Constructing partitioning table...")
    os.system(f"sudo parted -s -a optimal -- /dev/{b} mklabel msdos")
    time.sleep(0.5)
    print("partitioning 100% of volume to fat32...")
    os.system(f"sudo parted -s -a optimal -- /dev/{b} mkpart primary fat32 1MiB 100%")
    time.sleep(0.5)
    print("Cleaning up...")
    os.system(f"sudo parted -s -- /dev/{b} align-check optimal 1")
    time.sleep(0.5)
    print("Formatting Volume...")
    os.system(f"sudo mkfs.vfat -F32 /dev/{b}1")
    time.sleep(1)
    print("\nScript finished. Let's hope it worked. Remove SD card")
    return None

print("\nLaunching formatter2 program... ctrl+c to exit\n")
print("###############################################")

try: 
    while True:
        blocks1 = blockCheck()
        time.sleep(3)
        blocks2 = blockCheck()
        
        if len(blocks2) > len(blocks1): #Check if amount of volumes changes (aka. a block device is inserted)
            block = blocks2[0]
            print("CARD INSERTED. Contains the following:")
            os.system("lsblk")

            if input(f"Format volume /dev/{block}? (y/n): ") == "y":
                if "sd" in block:
                    wipeSD(block)
                elif "mmc" in block:
                    print("Sorry, can't interact with my own block (MMCBLK)")
                else:
                    print("Something went wrong - maybe the block isn't named 'sd...' ??")
            
            
        else:
            if waitstate == 1:  #Create an alternating wait messsage to make terminal look alive
                print("Waiting for SD....")
                waitstate = 0
            else:
                print("Waiting for SD..")
                waitstate = 1

except KeyboardInterrupt:
    print("\n\n\n...Exit by user!\n")

sderaser.py
# SDeraser live running script with button interaction (3.0).
# Searches for newly inserted volumes (blocks). Will attempt to delete the first block on list that starts with "sd..", following raspi block order ("sda, sdb, sdc.."" etc.).
# Uses a dual light push button as user interface. 
# Full documentation, and tutorial: https://github.com/airlabitu/SDeraser

# Thomas Kaufmanas // AIRLAB ITU // March 2023

import time
import os
import subprocess
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BOARD)
btnLED1 = 37
btnLED2 = 38
btn = 7
GPIO.setup(btnLED1,GPIO.OUT)
GPIO.setup(btnLED2,GPIO.OUT)
GPIO.setup(btn, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)

waitstate = 1

# SD functions to interact with block
def blockCheck():
    """Use subprocess module to read block devices inserted. Returns a list of names"""
    output = subprocess.Popen("lsblk -l -o NAME", shell=True, stdout=subprocess.PIPE).stdout
    output = output.read()
    output = output.decode("utf-8")
    output = output.strip().split("\n")
    result = output[1:]
    #print("array er:", result)
    return result
def wipeSD(blockname):
    """formats an SD card by name of primary blocks (thereby including any partitions under this)"""
    b = blockname
    print("Unmounting all partitions...")
    os.system(f"sudo umount /dev/{b}?")
    time.sleep(0.5)
    print("Constructing partitioning table...")
    os.system(f"sudo parted -s -a optimal -- /dev/{b} mklabel msdos")
    time.sleep(0.5)
    print("partitioning 100% of volume to fat32...")
    os.system(f"sudo parted -s -a optimal -- /dev/{b} mkpart primary fat32 1MiB 100%")
    time.sleep(0.5)
    print("Cleaning up...")
    os.system(f"sudo parted -s -- /dev/{b} align-check optimal 1")
    time.sleep(0.5)
    print("Formatting Volume...")
    os.system(f"sudo mkfs.vfat -F32 /dev/{b}1")
    time.sleep(1)
    print("\nScript finished. Let's hope it worked. Remove SD card")
    for i in range(2): #Button animation 
        btnRed()
        time.sleep(0.2)
        btnOff()
        time.sleep(0.2)
        i =+ 1
    btnOff()
    return None

# Button functions to switch polarity and thereby colour of button
def btnRed():
	"""Sets button colour to red"""
	GPIO.output(btnLED1, GPIO.HIGH)
	GPIO.output(btnLED2, GPIO.LOW)
def btnGreen():
	"""Sets button colour to green"""
	GPIO.output(btnLED1, GPIO.LOW)
	GPIO.output(btnLED2, GPIO.HIGH)
def btnOff():
	"""Sets button colour to off"""
	GPIO.output(btnLED1, GPIO.LOW)
	GPIO.output(btnLED2, GPIO.LOW)

GPIO.cleanup
btnOff()
print("\nResetting GPIO's and buttons, launching SDeraser program...\n")
print("###############################################")

# Main loop
try:
    while True:
        blocks1 = blockCheck()
        time.sleep(2)
        blocks2 = blockCheck()
        if len(blocks2) > len(blocks1):  #Check if amount of volumes changes (aka. a block device is inserted)
            block = blocks2[0]
            print("CARD INSERTED. Contains the following:")
            os.system("lsblk")
            btnGreen()
            print(f"Format volume /dev/{block}? \n Press green button to continue")

            while not GPIO.input(btn): #Loop holds the code until button is pushed to confirm the erase
                time.sleep(0.1)

            for i in range(3): #Button animation 
                btnRed()
                time.sleep(0.1)
                btnOff()
                time.sleep(0.1)
                i =+ 1
            btnRed()

            if "sd" in block:   # Only erase block if it is named SD (as is expected)
                wipeSD(block)            
            elif "mmc" in block:    # Reject if it tries to erase itself
                print("Sorry, can't interact with my own block (MMCBLK)")
            else:
                print("Something went wrong - maybe the block isn't named 'sd...?' ")
            
        else:
            if waitstate == 1:  #Create an alternating wait messsage to make terminal look alive
                print("Waiting for SD....")
                waitstate = 0
            else:
                print("Waiting for SD..")
                waitstate = 1

            

except KeyboardInterrupt:
    GPIO.cleanup()
    print("\n\n\n...Exit by user!\n")